packages/event-bus-redis/src/services/event-bus-redis.ts

```diff
@@ -0,0 +1,228 @@
+import {
+  AbstractEventBusModuleService,
+  ConfigModule,
+  ConfigurableModuleDeclaration,
+  Logger,
+  MODULE_RESOURCE_TYPE
+} from "@medusajs/medusa"
+import { Job, Queue, Worker } from "bullmq"
+import Redis from "ioredis"
+import { isDefined, MedusaError } from "medusa-core-utils"
+
+type InjectedDependencies = {
+  logger: Logger
+  configModule: ConfigModule
+}
+
+type EventBusRedisModuleOptions = {
+  queueName?: string
+  workerName?: string
+  queuePrefix?: string
+  workerPrefix?: string
+  redisUrl?: string
+}
+
+type BullJob<T> = {
+  data: {
+    eventName: string
+    data: T
+    completedSubscriberIds: string[] | undefined
+  }
+} & Job
+
+type EmitOptions = {
+  delay?: number
+  attempts: number
+  backoff?: {
+    type: "fixed" | "exponential"
+    delay: number
+  }
+}
+
+/**
+ * Can keep track of multiple subscribers to different events and run the
+ * subscribers when events happen. Events will run asynchronously.
+ */
+export default class RedisEventBusService extends AbstractEventBusModuleService {
+  protected readonly config_: ConfigModule
+  protected readonly moduleOptions_: EventBusRedisModuleOptions
+  protected readonly moduleDeclaration_: ConfigurableModuleDeclaration
+  protected readonly logger_: Logger
+
+  protected queue_: Queue
+  protected shouldEnqueuerRun: boolean
+  protected enqueue_: Promise<void>
+
+  constructor(
+    { logger, configModule }: InjectedDependencies,
+    moduleOptions: EventBusRedisModuleOptions = {},
+    moduleDeclaration: ConfigurableModuleDeclaration,
+    singleton = true
```

Comment:
todo: this can be removed. 

`singleton` is needed in the core to ensure that when the EventBus is cloned (e.g., in `withTransaction`) we don't again start the enqueuer (which reads from StagedJob and forwards into bull) otherwise we would potentially see that **one** staged job could result in **multiple** bull jobs.

Since this class doesn't do any "enqueuing work" we don't have to worry about that here